Add support for tintColor on ASImageNode and ASButtonNode

rahul-malik opened the pull request:
Adding the ability to specify a tintColor for images / buttons that behave similarly to how they would under UIKit

bolsinga commented on Source/ASButtonNode.mm:
```diff
@@ -55,7 +55,7 @@ - (ASTextNode *)titleNode
       // tvOS needs access to the underlying view
       // of the button node to add a touch handler.
-    [_titleNode setLayerBacked:YES];
+//    [_titleNode setLayerBacked:YES];
```
Why is this commented out?

bolsinga commented on Source/ASButtonNode.mm:
```diff
@@ -69,7 +69,7 @@ - (ASImageNode *)imageNode
   if (!_imageNode) {
     _imageNode = [[ASImageNode alloc] init];
-    [_imageNode setLayerBacked:YES];
+ //   [_imageNode setLayerBacked:YES];
```
Why is this commented out?

rahul-malik replied:
You can't set the tintColor of a layer-backed node without hitting an assertion currently

rahul-malik commented on Source/ASTextNode.mm:
```diff
@@ -570,7 +575,13 @@ + (UIImage *)displayWithParameters:(id<NSObject>)parameters isCancelled:(NS_NOES
     }
-    
+
+    // Set stroke path if tint is specified
```
Note: Will remove this since it's not needed anymore

rahul-malik commented on Tests/ASButtonSnapshotTests.mm:
```diff
@@ -0,0 +1,58 @@
+
+
+- (void)testTintColor
+{git
```
esh, git should not be in here :)

rahul-malik commented on Tests/ASImageNodeSnapshotTests.mm:
```diff
@@ -72,7 +71,43 @@ - (void)testTintColorBlock
+  UIImage *test = [self testImage];
+  ASImageNode *node = [[ASImageNode alloc] init];
+  node.image = [test imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
+  // Tint color should not have
```
I'll finish this thought

bolsinga commented on Source/ASTextNode.mm:
```diff
@@ -570,7 +575,13 @@ + (UIImage *)displayWithParameters:(id<NSObject>)parameters isCancelled:(NS_NOES
+
+    // Set stroke path if tint is specified
+    if (tintColor) {
+      [tintColor setStroke]; // doesn't work? overwrite attributedString here?
```
Can you elaborate on this comment?

rahul-malik replied:
Reverted

rahul-malik commented on Source/ASTextNode.mm:
```diff
@@ -136,6 +136,7 @@ @interface ASTextNodeDrawParameter : NSObject {
 @package
   ASTextKitAttributes _rendererAttributes;
   UIColor *_backgroundColor;
+  UIColor *_tintColor;
```
I think we'll actually undo all the changes to this file since the "tint" is handled from ASButtonNode right now

bolsinga commented on Tests/ASButtonSnapshotTests.mm:
```diff
@@ -0,0 +1,58 @@
+
+
+- (void)testTintColor
+{git
```
typo?

rahul-malik replied:
Fixed!

rcancro commented on Source/ASButtonNode.mm:
```diff
@@ -301,12 +350,17 @@ - (void)setImageAlignment:(ASButtonNodeImageAlignment)imageAlignment
+    attributes[NSForegroundColorAttributeName] = color;
+  } else if (self.tintColor != nil) {
+    attributes[NSForegroundColorAttributeName] = self.tintColor;
+  }
```
Do we need a fallback like before in the case that `color` and `tintColor` are nil?

rahul-malik replied:
I don't believe so, if you look at the snapshot tests you'll see that when we specify `nil` as the background color it ends up being black anyways which preserves the existing logic

rcancro commented on Source/ASButtonNode.mm:
```diff
@@ -131,6 +131,55 @@ - (void)setDisplaysAsynchronously:(BOOL)displaysAsynchronously
+{
+  if (title) {
+    NSMutableAttributedString *mutString = [[NSMutableAttributedString alloc] initWithAttributedString:title];
+    NSRange limit = NSMakeRange(0, _normalAttributedTitle.length);
```
should this be `title`, not `_normalAttributedTitle`? It looks like several different titles are being passed in here in `updateTitleForegroundColor`

rahul-malik replied:
Great catch!

rcancro commented on Source/ASButtonNode.mm:
```diff
@@ -131,6 +131,55 @@ - (void)setDisplaysAsynchronously:(BOOL)displaysAsynchronously
+    NSRange limit = NSMakeRange(0, _normalAttributedTitle.length);
+    NSRange effectiveRange;
+    UIColor *attributeValue = (UIColor *)[mutString attribute:NSForegroundColorAttributeName atIndex:limit.location effectiveRange:&effectiveRange];
+    if (attributeValue == nil) {
```
Does this mean we are only setting the foreground color if one was set previously?

rahul-malik replied:
We're only setting the tintColor as the titleColor if there was not one set previously

rahul-malik replied:
I think this method name is confusing actually, will update

pinterest-ios commented:
<!--
  0 Errors
  1 Warning: Please ensure license is corre...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please ensure license is correct for ASButtonSnapshotTests.mm:</p>
<pre><code>//
//  ASButtonSnapshotTests.mm
//  Texture
//
//  Copyright (c) Pinterest, Inc.  All rights reserved.
//  Licensed under Apache 2.0: http://www.apache.org/licenses/LICENSE-2.0
</code></</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/202971b7ac75038283bbe6eb20936f59/raw/35b471f9886c2dcedb2c596ab8000c1e7b5b524a/log.txt)

pinterest-ios commented:
<!--
  0 Errors
  1 Warning: Please ensure license is corre...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please ensure license is correct for ASButtonSnapshotTests.mm:</p>
<pre><code>//
//  ASButtonSnapshotTests.mm
//  Texture
//
//  Copyright (c) Pinterest, Inc.  All rights reserved.
//  Licensed under Apache 2.0: http://www.apache.org/licenses/LICENSE-2.0
</code></</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

pinterest-ios commented:
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/295e52cf858f68961ec160f36ac6c010/raw/aea8b275906ab670b802bcfca87036d2a6eaeb95/log.txt)

rahul-malik commented:
@nguyenhuy @bolsinga - Please take another look. 

I've reworked the PR in a few important ways:
- Tinting for ASTextNode is now handled at the ASTextKitRenderer level
- More snapshot tests to make sure changing the tint color works correctly
- Migrate logic in ASButtonNode from `setTintColor` to `tintColorDidChange`. We can rely on `UIView` to call that method instead of overriding the setter
- Tinting in ASImageNode is now done inline. Code is duplicated but is fairly small and allows us to use the theoretically more performant code path (still needs measurement)
- Updated headers of new files for proper licensing

pinterest-ios commented:
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/f5cff2deb15944314ec0bc1e0adc470b/raw/fb4d1d7223d7e15600a9a9260f3804c6bbecba21/log.txt)

pinterest-ios commented:
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/6456dc5b196c23c8c96a56bd67876069/raw/6c9cb5691430656dda55b7e4106b26b104797490/log.txt)

pinterest-ios commented:
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/9a60b81681dea70b8f25e3d80ea5a8d0/raw/664f5103767992d358dcfa4ead133704dc69faae/log.txt)

pinterest-ios commented:
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/d7e3a816bfd1c658ce17baa002e3cc05/raw/f8f155aacffe156c8d85ec0b7929262a1571fa06/log.txt)

pinterest-ios commented:
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/19b6c1b0dd707ec791757f681e4efeeb/raw/da9437363d0c4bc40745f9ffa3d3331fecb80295/log.txt)

nguyenhuy commented:
To other reviewers: I'm going to land this diff now so we can start testing it internally (we have a fixed sync schedule now). Feel free to do a post-review, especially the changes in ASTextNode2, and we'll follow up on any issues found.

nguyenhuy commented on Source/ASImageNode.mm:
```diff
@@ -499,6 +507,11 @@ + (UIImage *)createContentsForkey:(ASImageNodeContentsKey *)key drawParameters:(
+    UIImageRenderingMode renderingMode = [image renderingMode];
+    if (renderingMode == UIImageRenderingModeAlwaysTemplate && key.tintColor) {
+      [key.tintColor setFill];
+      image = [image imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
```
```suggestion
```
